### PR TITLE
Auto-detect engine when restoring backups

### PR DESCRIPTION
**Pull request by enisoc**
The idea is to make it easier to change the backup engine on existing clusters. The flag now only controls the engine used to take new backups. When restoring, the flag will be ignored since we can tell which engine to use by looking at the MANIFEST.

That means you can change the engine and roll out the flag change without worrying that tablets will be unable to restore until a new backup is taken with the new engine.

Signed-off-by: Anthony Yeh <enisoc@planetscale.com>

### CODE REVIEW

File: go/vt/mysqlctl/backupengine.go [hunk 1]
```diff
@@ -34,47 +34,121 @@ import (
...
 
-// GetBackupEngine returns the current BackupEngine implementation.
-// Should be called after flags have been initialized.
+// BackupEngineMap contains the registered implementations for BackupEngine
+var BackupEngineMap = make(map[string]BackupRestoreEngine)
+
+// GetBackupEngine returns the BackupEngine implementation that should be used
+// to create new backups.
+//
+// To restore a backup, you should instead get the appropriate BackupEngine for
```
Comment (deepthi) [inline go/vt/mysqlctl/backupengine.go]: BackupEngine -> RestoreEngine

File: go/vt/mysqlctl/backupengine.go [hunk 2]
```diff
@@ -34,47 +34,121 @@ import (
...
+// BackupRestoreEngine is a combination of BackupEngine and RestoreEngine.
+type BackupRestoreEngine interface {
+	BackupEngine
+	RestoreEngine
+}
 
-// GetBackupEngine returns the current BackupEngine implementation.
-// Should be called after flags have been initialized.
+// BackupEngineMap contains the registered implementations for BackupEngine
+var BackupEngineMap = make(map[string]BackupRestoreEngine)
```
Comment (deepthi) [inline go/vt/mysqlctl/backupengine.go]: I suggest renaming this to BackupRestoreEngineMap to reflect its new purpose. Also fix doc comment.

File: go/vt/mysqlctl/backupengine.go [hunk 3]
```diff
@@ -34,47 +34,121 @@ import (
...
+// a particular backup by calling GetRestoreEngine().
+//
+// This must only be called after flags have been parsed.
 func GetBackupEngine() (BackupEngine, error) {
-	be, ok := BackupEngineMap[*backupEngineImplementation]
+	name := *backupEngineImplementation
+	be, ok := BackupEngineMap[name]
 	if !ok {
-		return nil, vterrors.New(vtrpc.Code_NOT_FOUND, "no registered implementation of BackupEngine")
+		return nil, vterrors.Errorf(vtrpc.Code_NOT_FOUND, "unknown BackupEngine implementation %q", name)
```
Comment (deepthi) [inline go/vt/mysqlctl/backupengine.go]: +1

**Comment by rafael** [comment]
Oh this improvement is awesome. We were actually thinking about this for our migration to xtrabackups.

**Pull request merged**